sdk/identity/azure-identity/test/ut/default_azure_credential_test.cpp

```diff
@@ -354,3 +354,171 @@ TEST_P(LogMessages, )
+        SpecificCredentialInfo{"WorkloadIdentityCredential", "workloadidentitycredential", 4},
+        SpecificCredentialInfo{"ManagedIdentityCredential", "MANAGEDIDENTITYCREDENTIAL", 8},
+        SpecificCredentialInfo{"AzureCliCredential", "  AzureCLICredential ", 4}));
+
+TEST_P(LogMessagesForSpecificCredential, )
+{
+  const auto specificCredentialInfo = GetParam();
+
+  using LogMsgVec = std::vector<std::pair<Logger::Level, std::string>>;
+  LogMsgVec log;
+  Logger::SetLevel(Logger::Level::Verbose);
+  Logger::SetListener([&](auto lvl, auto msg) { log.push_back(std::make_pair(lvl, msg)); });
+
+  try
+  {
+    CredentialTestHelper::EnvironmentOverride const env({
+        {"AZURE_TENANT_ID", "01234567-89ab-cdef-fedc-ba8976543210"},
+        {"AZURE_CLIENT_ID", "fedcba98-7654-3210-0123-456789abcdef"},
+        {"AZURE_CLIENT_SECRET", "CLIENTSECRET"},
+        {"AZURE_AUTHORITY_HOST", "https://microsoft.com/"},
+        {"AZURE_FEDERATED_TOKEN_FILE", "azure-identity-test.pem"},
+        {"AZURE_USERNAME", ""},
+        {"AZURE_PASSWORD", ""},
+        {"AZURE_CLIENT_CERTIFICATE_PATH", ""},
+        {"MSI_ENDPOINT", ""},
+        {"MSI_SECRET", ""},
+        {"IDENTITY_ENDPOINT", ""},
+        {"IMDS_ENDPOINT", ""},
+        {"IDENTITY_HEADER", ""},
+        {"IDENTITY_SERVER_THUMBPRINT", ""},
+        {"AZURE_TOKEN_CREDENTIALS", specificCredentialInfo.EnvVarValue},
+    });
+
+    static_cast<void>(std::make_unique<DefaultAzureCredential>());
+
+    EXPECT_EQ(log.size(), specificCredentialInfo.ExpectedLogMsgCount);
+    LogMsgVec::size_type i = 0;
+
+    EXPECT_EQ(log.at(i).first, Logger::Level::Verbose);
+    EXPECT_EQ(
+        log.at(i).second,
+        "Identity: Creating DefaultAzureCredential which combines "
+        "multiple parameterless credentials into a single one."
+        "\nDefaultAzureCredential is only recommended for the early stages of development, "
+        "and not for usage in production environment."
+        "\nOnce the developer focuses on the Credentials "
+        "and Authentication aspects of their application, "
+        "DefaultAzureCredential needs to be replaced with the credential that "
+        "is the better fit for the application.");
+
+    ++i;
+    EXPECT_EQ(log.at(i).first, Logger::Level::Verbose);
+    EXPECT_EQ(
+        log.at(i).second,
+        "Identity: DefaultAzureCredential: "
+        "'AZURE_TOKEN_CREDENTIALS' environment variable is set to '"
+            + specificCredentialInfo.EnvVarValue
+            + "', therefore credential chain will only contain single credential: "
+            + specificCredentialInfo.CredentialName + '.');
+
+    if (specificCredentialInfo.CredentialName == "EnvironmentCredential")
+    {
+      ++i;
+      EXPECT_EQ(log.at(i).first, Logger::Level::Informational);
+      EXPECT_EQ(
+          log.at(i).second,
+          "Identity: EnvironmentCredential gets created with ClientSecretCredential.");
```

Comment:
Is it feasible to test the composition of the chain directly instead of using log messages as a proxy?

Reply:
I don't see an easy way of doing it. The credential is encapsulated behind the Chained/DAC credential implementation, it's goal is to make the credential feel as a whole credential, I don't see an easy way to get inside of its implementation from the outside and check which credential is being used.
This does not only test which credential is being used, but also check the quality of logs, which for Identity and specifically DAC/Chained Cred we consider a feature in itself - not just some debug records of varying quality, but it is an interface for the user to look into what is happening, it is part of the value provided by the credential, the log should be easy to read.